### PR TITLE
Tutorial for the External Contacts Scanning API

### PR DESCRIPTION
**Pull request by themooks**
Tutorial for the External Contacts Scanning API

### CODE REVIEW

File: scan-external-contacts/config.yml [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+    Go here for more information about [External Contacts Scanning](/api/rest/v2/externalcontacts/scan.html).
+
+    For this tutorial, we are going to use the Genesys Cloud SDK, and we will scan and export the External Contacts.
+    
+    The sample code is using javascript async/await (making it synchronous) to ease readibility. You can change and update it based on your needs.
+
+- title: Authentication
+  content: |
+    You will need an OAuth client using Client Credentials to authenticate. See [Create an OAuth Client](https://help.mypurecloud.com/articles/create-an-oauth-client/) for more info
```
Comment (PrinceNebulon) [inline scan-external-contacts/config.yml:21]: This makes it sound like these APIs can only be used with client credentials. Assuming that's not the case, can you reword to be more generic so the statement works for any client type? Or be more explicit that the tutorial is using client credentials, but you can use user-based auth as well?

File: scan-external-contacts/config.yml [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    
+    You can adapt the pause duration to your needs and current use of API in your organization.
+    
+    As the scan of External Contacts (for export/save) is likely not something that has to return immediately (no real-time needs), it shouldn't be a problem to make a pause, and even to execute your code via a nightly batch.
+
+    The sample also manages a maximum number of consecutive login failures or rate limit errors (the application will exit).
+
+    It is possible to export the External Contacts to a CSV file (and to configure the field separator) or to a raw JSON file (JSON Array of External Contacts).
+
+- title: Main Function
```
Comment (PrinceNebulon) [inline scan-external-contacts/config.yml:51]: Titles should generally be unique. If there's nothing else to call these steps, maybe combine them.

File: scan-external-contacts/nodejs/index.js [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+    }
+    catch (error) {
+        console.error('Login error - Error:', JSON.stringify(error, null, 2));
+
+        return false;
+    }
+}
+
+// Perform an External Contacts Scan API request and manage errors (401, 429)
+const gcloudContactsScan = async (cursor) => {
```
Comment (PrinceNebulon) [inline scan-external-contacts/nodejs/index.js:63]: "gcloud" isn't an acceptable shortening of Genesys Cloud; use the full name or avoid it altogether. (per marketing, don't kill the messenger)

File: scan-external-contacts/nodejs/index.js [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+                    // Try to request the same data again (current value of cursor)
+                    let retryScanData = await gcloudContactsScan(cursor);
+                    return retryScanData;
+                } else {
+                    result.errorMsg = "Unsuccessful Relogin attempt. Exiting...";
+                    result.error = true;
+                    return result;
+                }
+            }
+        } else if (error.status && error.status === 429) {
```
Comment (PrinceNebulon) [inline scan-external-contacts/nodejs/index.js:136]: The JS SDK will have built-in rate limit retrying when DEVENGAGE-271 is implemented. Preferably, remove all of the logic related to this. If this tutorial can't function without handling rate limiting, please implement rate limiting logic per the documented best practices: https://developer.mypurecloud.com/api/rest/rate_limits.html#best_practices_for_retrying_requests

File: scan-external-contacts/nodejs/index.js [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        // Reset of some global vars - counters for retry
+        consecutiveLoginRetryAttempts = 0;
+        consecutiveRateLimitRetryAttempts = 0;
+
+        return result;
+    }
+    catch (error) {
+        console.log('External Contacts Scan error - Error:', JSON.stringify(error, null, 2));
+
+        if (error.status && error.status === 401) {
```
Comment (PrinceNebulon) [inline scan-external-contacts/nodejs/index.js:113]: Access tokens shouldn't be getting invalidated during normal execution of a tutorial. Please remove this logic in favor of keeping tutorials as succinct as possible. Tutorials are intended to briefly demonstrate how to orchestrate API requests to accomplish a task; they are not meant to implement best practices for writing apps for production.

File: scan-external-contacts/nodejs/index.js [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+                    // Note: filter - You can implement a filter condition here and decide if the entity is to be printed (proceed) or not (return)
+                    // Note: transform - You can manipulate the entity attributes to create the desired string to be printed
+
+                    if (exportType === 'CSV') {
+                        // CSV
+                        if (writeStream) {
+                            for (let entity of scanResult.entities) {
+                                // For each entity element, print a line (CSV) with a field separator between values
+                                let entityString = (entity.id ? entity.id : '') + fieldSeparator + (entity.modifyDate ? entity.modifyDate : '') + fieldSeparator + (entity.firstName ? entity.firstName : '') + fieldSeparator + (entity.lastName ? entity.lastName : '') + fieldSeparator + (entity.workPhone ? entity.workPhone.e164 : '') + fieldSeparator + (entity.workEmail ? entity.workEmail : '') + fieldSeparator + (entity.externalOrganization ? entity.externalOrganization.id : '') + fieldSeparator + (entity.surveyOptOut ? entity.surveyOptOut.toString() : 'false') + "\n";
```
Comment (PrinceNebulon) [inline scan-external-contacts/nodejs/index.js:246]: Don't build CSV files manually. Use a [node library for CSV handling](https://www.npmjs.com/search?q=csv) to ensure proper field escaping and file formatting.

**Comment by themooks** [comment]
Update according to feedback/comments

File: scan-external-contacts/config.yml (hunk 1 above)
Comment (carnellj) [reply scan-external-contacts/config.yml]: One thing you might want to note too is that too many people use client credential grants because they are the easiest to understand.  The reality is that if you are going to be calling this as a service from a web facing application you should be using a client auth.

File: scan-external-contacts/nodejs/index.js (hunk 3 above)
Comment (carnellj) [reply scan-external-contacts/nodejs/index.js]: I wouldnt even include the name in the variable.  contactsScan is pretty descriptive.

File: scan-external-contacts/nodejs/index.js (hunk 4 above)
Comment (carnellj) [reply scan-external-contacts/nodejs/index.js]: Nice.

File: scan-external-contacts/config.yml [hunk 7]
```diff
@@ -0,0 +1,144 @@
+---
+title: Scan through External Contacts
+description: See how to scan through external contacts, external organizations, relationships and notes using the External Contacts API.
+categories:
+- Contacts
+steps:
+- title: Introduction
+  content: |
+    This tutorial walks through the steps of implementing a function to scan through the external contacts of a Genesys Cloud organization, and to process them before saving them into a file.
```
Comment (carnellj) [inline scan-external-contacts/config.yml:9]: It might be useful to introduce what external contact is and why you would be interested in using them.  Sometimes I think our tutorials assume a certain level of familiarity with the topics, but you might be dealing with someone completely new and that little bit fo background would help the reader.

File: scan-external-contacts/config.yml [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+- title: Scan in loop
+  content: |
+    The main function will then initiate a loop, [scanning for External Contacts page per page](/api/rest/v2/externalcontacts/#get-api-v2-externalcontacts-scan-contacts), until all data is retrieved.
+
+- title: Process and Export retrieved data
+  content: |
+    If a page containing External Contacts is retrieved, they are processed and written to the output file (via the Stream) according to the type of export (CSV or raw JSON).
+
+    Note that an External Contacts Scan request can be successful while not bringing back any External Contacts (entities array attribute is empty).
```
Comment (carnellj) [inline scan-external-contacts/config.yml:72]: This seems a little hard to follow. Shouldn't the HTTP status code provide a response to what happened?

File: scan-external-contacts/nodejs/index.js [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+
+        // Reset of some global vars - counters for retry
+        consecutiveRateLimitRetryAttempts = 0;
+
+        return result;
+    }
+    catch (error) {
+        console.log('External Contacts Scan error - Error:', JSON.stringify(error, null, 2));
+
+        if (error.status && error.status === 429) {
```
Comment (carnellj) [inline scan-external-contacts/nodejs/index.js:111]: The JS SDK has automatic retry logic on 429's.  This would be a great place to show case this.

File: scan-external-contacts/nodejs/index.js [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+                let retryAfter = 0;
+                if (error.headers && error.headers.hasOwnProperty("retry-after")) {
+                    retryAfter = parseInt(error.headers["retry-after"]);
+                } else {
+                    result.errorMsg = "Failed to retrieve Retry-After. Exiting...";
+                    result.error = true;
+                    return result;
+                }
+                // Add 1 second (for "safety")
+                let retryIn = (retryAfter + 1) * 1000;
```
Comment (carnellj) [inline scan-external-contacts/nodejs/index.js:129]: Rather than writing this yourself, you could steer the reader to using a library like lifeomic's attempt library.  I used it in the user-provisioning-scripts developer guide and might be a better solution than rolling our own.  (https://github.com/lifeomic/attempt)

File: scan-external-contacts/nodejs/index.js [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+
+/*
+------------------------------------------------------------------------------
+* Main
+------------------------------------------------------------------------------
+*/
+
+(async () => {
+
+    if (maxRecordsPerPage < 10 || maxRecordsPerPage > 200) {
```
Comment (carnellj) [inline scan-external-contacts/nodejs/index.js:161]: This method is really too big.  I like to break my "main" method into a series of function calls that simply step through each step.  I would break this funtion into smaller pieces and then walk through each function.

File: scan-external-contacts/nodejs/index.js [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+        let opts = {
+            'limit': maxRecordsPerPage
+        };
+        // No opts.cursor on first scan - cursor == null
+        if (cursor) {
+            opts.cursor = cursor;
+        }
+
+        // External Contacts Scan API request
+        let scanData = await contactsApi.getExternalcontactsScanContacts(opts);
```
Comment (carnellj) [inline scan-external-contacts/nodejs/index.js:87]: I would recommend you take a look at https://github.com/MyPureCloud/user-provisioning-scripts-js/blob/master/src/proxies/groupsapi.js.  I have got a pretty good pattern for how to assemble pages of data into a single data structure.  The key thing is really understanding how much data you are bringing back.  If it massive, you will run of memory, but most calls are going to be small in size with a few pages of data being retrieved.  The code tends to be more readable and then you can process the data as a single flate data structure.  Your approach works but it is very verbose.

File: scan-external-contacts/nodejs/index.js [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+
+                let elapsedTime = Date.now() - scanResult.startTime;
+                console.log("Execution of scan - Start At: " + (scanResult.startTime).toString() + " - Duration: " + elapsedTime.toString() + " ms");
+
+                if (scanResult.isFinished === true) {
+                    // Last Data was received (no nextCursor) and export is finished
+                    isCompleted = true;
+                } else {
+                    // Force a pause before next scan (to avoid loading org/token unnecessarily - API rate limit)
+                    let forcedDelay;
```
Comment (carnellj) [inline scan-external-contacts/nodejs/index.js:247]: This type of logic is better encapsulated in using a retry code with exponential backoff.  You are trying to show how to use the API not getting buried in the details of pagination logic and error handling :)

**Review by carnellj — COMMENTED** [review]
Overall pretty good.  Thanks for putting this together.  If  you want to chat more about my feedback please reach out to me.  My way is not the only way, but I just got done righting a lot of Javascript and got to spend a lot of time with people reviewing my Javascript code.

File: scan-external-contacts/nodejs/index.js (hunk 9 above)
Comment (PrinceNebulon) [reply scan-external-contacts/nodejs/index.js:111]: It doesn't currently. DEVENGAGE-271

**Comment by themooks** [comment]
I have tried to take into account the different feedback. Hope I haven't made a step backwards :-)

**Review by timsmithgenesys — APPROVED** [review]
